librad/src/git/storage.rs

```diff
@@ -136,8 +136,65 @@ impl Storage {
             .or_matches(is_not_found_err, || Ok(false))
     }
 
-    #[tracing::instrument(level = "debug", skip(self), err)]
+    /// Check the existence of `oid` as a **commit**.
+    ///
+    /// The result will be `false` if:
+    ///
+    /// 1. No commit could be found for `oid`
+    /// 2. The reference path for the `urn` could not be found (it defaults to
+    /// `rad/id` if not provided)
+    /// 3. The tip SHA was not in the history of the commit
+    /// 4. The `oid` was the [`zero`][`git2::Oid::zero`] SHA.
+    #[tracing::instrument(level = "debug", skip(self, urn), fields(urn = %urn), err)]
     pub fn has_commit<Oid>(&self, urn: &Urn, oid: Oid) -> Result<bool, Error>
+    where
+        Oid: AsRef<git2::Oid> + Debug,
+    {
+        let oid = match self.find_object(oid)? {
+            None => return Ok(false),
+            Some(object) => match object.kind() {
+                Some(git2::ObjectType::Commit) => object.id(),
+                _ => return Ok(false),
+            },
+        };
+
+        let tip = self.tip(urn)?;
+        Ok(tip
+            .map(|tip| {
+                Ok::<_, git2::Error>(*tip == oid || self.backend.graph_descendant_of(*tip, oid)?)
+            })
+            .transpose()?
+            .unwrap_or(false))
+    }
+
+    /// Check the existence of `oid` as a **tag**.
+    ///
+    /// The result will be `false` if:
+    ///
+    /// 1. No tag could be found for `oid`
+    /// 2. The reference path for the `urn` could not be found (it defaults to
+    /// `rad/id` if not provided)
+    /// 3. The SHA of the tag was not the same as the resolved reference
+    /// 4. The `oid` was the [`zero`][`git2::Oid::zero`] SHA.
+    #[tracing::instrument(level = "debug", skip(self, urn), fields(urn = %urn), err)]
+    pub fn has_tag<Oid>(&self, urn: &Urn, oid: Oid) -> Result<bool, Error>
+    where
+        Oid: AsRef<git2::Oid> + Debug,
+    {
+        let oid = match self.find_object(oid)? {
+            None => return Ok(false),
+            Some(object) => match object.kind() {
+                Some(git2::ObjectType::Tag) => object.id(),
+                _ => return Ok(false),
+            },
+        };
+
+        let tip = self.tip(urn)?;
```

Comment:
Uh-oh wait -- `tip` peels to a commit, which is _probably_ not the `oid` passed here. Just using `refname_to_id` here is cheaper and what you want.

I'm not sure if we should remove the implicit type conversion of `tip`. I think this would break the assumptions of the identities stuff, so perhaps it should return `Object` instead. This puts the burden of deciding what to do if the tip is _not_ a commit on the caller, which may require some thought.

Reply:
Ach nooooo. I completely missed that. Should have got better laser eye surgery.

> I think this would break the assumptions of the identities stuff, so perhaps it should return Object instead. This puts the burden of deciding what to do if the tip is not a commit on the caller, which may require some thought.

I tried grepping for its use and it seems to only be used here. The identity stuff seems to use `peel_to_commit` directly. So the only affected spots would be `has_commit` and `has_tag` here.